File: 491-non-decreasing-subsequences/non-decreasing-subsequences.py
class Solution(object):
    def findSubsequences(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        result = set()
        
        def solve(prev, i, combi):
            if i == len(nums):
                if len(combi) >= 2:
                    result.add(tuple(combi))  # Convert to tuple to store in set
                return

            if nums[i] >= prev:
                combi.append(nums[i])
                solve(nums[i], i + 1, combi)
                combi.pop()  # Backtrack

            # Also part of backtrack - skip current element
            solve(prev, i + 1, combi)
        
        solve(-101, 0, [])
        return [list(seq) for seq in result]  # Convert back to list

File: 491-non-decreasing-subsequences/non-decreasing-subsequences.cpp
class Solution {
    void solve(set<vector<int>>& result, vector<int>& nums, int prev, int i, vector<int>& combi) {
        // Base condition
        if (i == nums.size()) {
            if (combi.size() >= 2) {
                result.insert(combi);
            }
            return;
        }

        if (nums[i] >= prev) {
            combi.push_back(nums[i]);
            solve(result, nums, nums[i], i + 1, combi);

            // Backtrack
            combi.pop_back();
        }

        // Also part of backtrack - skip current element
        solve(result, nums, prev, i + 1, combi);
    }

public:
    vector<vector<int>> findSubsequences(vector<int>& nums) {
        set<vector<int>> result;
        vector<int> combi;

        solve(result, nums, -101, 0, combi);

        return vector<vector<int>>(result.begin(), result.end());
    }
};
